CMakeLists.txt

```diff
@@ -133,13 +134,23 @@ add_subdirectory(${xgboost_SOURCE_DIR}/plugin)
 add_subdirectory(${xgboost_SOURCE_DIR}/src)
 set(XGBOOST_OBJ_SOURCES "${XGBOOST_OBJ_SOURCES};$<TARGET_OBJECTS:objxgboost>")
 
-#-- Shared library
-add_library(xgboost SHARED ${XGBOOST_OBJ_SOURCES})
-target_include_directories(xgboost
-  INTERFACE
-  $<INSTALL_INTERFACE:${CMAKE_INSTALL_PREFIX}/include>
-  $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>)
-target_link_libraries(xgboost PRIVATE ${LINKED_LIBRARIES_PRIVATE})
+if (BUILD_STATIC_LIB)
+    #-- Static library
+    add_library(xgboost STATIC ${XGBOOST_OBJ_SOURCES})
```

Comment:
Can you replace this entire if statement with a cmake generator expression to determine the STATIC/SHARED keyword?

Reply:
I tried using generator expressions but couldn't make it work correctly with an cmake option. Another way is to drop setting the library type explicitly and to use the cmake property https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html. The user can then build statically with ` -DBUILD_SHARED_LIBS=OFF `.

One issue that I did find is the library does not seem to obey `set_output_directory` and leaves the compiles `libxgboost.a` file in the top level build directory. This was true for both attempts at this.

Reply:
Add the following to the function `set_output_directory` in cmake/Utils.cmake:
```
		ARCHIVE_OUTPUT_DIRECTORY ${dir}
		ARCHIVE_OUTPUT_DIRECTORY_DEBUG ${dir}
		ARCHIVE_OUTPUT_DIRECTORY_RELEASE ${dir}
		ARCHIVE_OUTPUT_DIRECTORY_RELWITHDEBINFO ${dir}
		ARCHIVE_OUTPUT_DIRECTORY_MINSIZEREL ${dir}
```
This will place the static library in the correct place.

Reply:
And yes exposing `BUILD_SHARED_LIBS` as an option seems to be the correct way to do this.